### ISSUE TITLE
Proposal: Add Socket.SafeHandle property

### ISSUE THREAD

**Issue by stephentoub** [dotnet/corefx#13470]
Socket currently exposes:
```C#
public IntPtr Handle => _handle.DangerousGetHandle();
```

We should add:
```C#
public SafeHandle SafeHandle => _handle;
```

This will let code consuming the socket interact with its handle directly in a safe manner; today if code relies on the Handle, the Socket could be closed out from under it, closing that handle, and leading to dangerous race conditions.  With the SafeHandle exposed, consuming code can handle this correctly.

For example, this will make it feasible to correctly fix https://github.com/dotnet/corefx/issues/11807, whereas today reflection is used:
https://github.com/dotnet/corefx/blob/master/src/System.IO.Pipes/src/Microsoft/Win32/SafeHandles/SafePipeHandle.Unix.cs#L36

---
Updated by @stephentoub to reflect API review outcome:
```C#
public SafeSocketHandle SafeHandle  { get { ... } }
...
public class SafeSocketHandle : SafeHandleMinusOneIsInvalid
{
    public SafeSocketHandle(IntPtr preexistingHandle, bool ownsHandle);
}
```

**Comment by davidsh** [comment]
cc: @CIPop

**Comment by terrajobst** [comment]
We should probably add a strongly typed handle class, like `SafeSocketHandle` as we can't easily do this later. Other then that, it looks good.

**Comment by danmosemsft** [comment]
@stephentoub are you planning to do this? Perhas not & we can put it up for grabs

**Comment by stephentoub** [comment]
@danmosemsft, up for grabs is fine.  If I can get to it, I can always grab it.

**Comment by davidsh** [comment]
requires API change to System.Net.Sockets contract.  New API won't be part of .NETStandard2.0.

**Comment by karelz** [comment]
Moving to Future. It should not block 2.0 release.

**Comment by karelz** [comment]
Fixed in PR #32649 (not marked as merged due to GitHub hiccups)
